### PR TITLE
KAFKA-2607: Review `Time` interface and its usage

### PR DESCRIPTION
**Pull request by afine**
This change adds some documentation and changes some calls from using absolute time to relative time.

Since I am new to the Kafka codebase I will mark the locations where I made the change from absolute to relative times to make it easier for reviewers to find any bugs that I might have introduced.

### CODE REVIEW

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 1]
```diff
@@ -418,7 +418,7 @@ class Partition(val topic: String,
     val leaderLogEndOffset = leaderReplica.logEndOffset
     val candidateReplicas = inSyncReplicas - leaderReplica
 
-    val laggingReplicas = candidateReplicas.filter(r => (time.milliseconds - r.lastCaughtUpTimeMs) > maxLagMs)
```
Comment (afine) [inline core/src/main/scala/kafka/cluster/Partition.scala:421]: changed from absolute to relative time

File: core/src/main/scala/kafka/cluster/Replica.scala [hunk 2]
```diff
@@ -45,7 +45,7 @@ class Replica(val brokerId: Int,
     }
   }
 
-  private[this] val lastCaughtUpTimeMsUnderlying = new AtomicLong(time.milliseconds)
```
Comment (afine) [inline core/src/main/scala/kafka/cluster/Replica.scala:48]: changed from absolute to relative time

File: core/src/main/scala/kafka/cluster/Replica.scala [hunk 3]
```diff
@@ -57,7 +57,7 @@ class Replica(val brokerId: Int,
      * This means that the replica is fully caught up.
      */
     if(logReadResult.isReadFromLogEnd) {
-      lastCaughtUpTimeMsUnderlying.set(time.milliseconds)
```
Comment (afine) [inline core/src/main/scala/kafka/cluster/Replica.scala:60]: changed from absolute to relative time

File: core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala [hunk 4]
```diff
@@ -358,7 +358,7 @@ class GroupMetadataManager(val brokerId: Int,
         }
       }
 
-      val startMs = SystemTime.milliseconds
```
Comment (afine) [inline core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:361]: changed from absolute to relative time

File: core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala [hunk 5]
```diff
@@ -437,7 +437,7 @@ class GroupMetadataManager(val brokerId: Int,
 
             if (!shuttingDown.get())
               info("Finished loading offsets from %s in %d milliseconds."
-                .format(topicPartition, SystemTime.milliseconds - startMs))
```
Comment (afine) [inline core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:440]: changed from absolute to relative time

File: core/src/main/scala/kafka/log/LogCleaner.scala [hunk 6]
```diff
@@ -659,11 +659,11 @@ private case class CleanerStats(time: Time = SystemTime) {
   }
 
   def indexDone() {
-    mapCompleteTime = time.milliseconds
```
Comment (afine) [inline core/src/main/scala/kafka/log/LogCleaner.scala:662]: changed from absolute to relative time

File: core/src/main/scala/kafka/log/LogCleaner.scala [hunk 7]
```diff
@@ -659,11 +659,11 @@ private case class CleanerStats(time: Time = SystemTime) {
...
+    mapCompleteTime = time.relativeMilliseconds
   }
 
   def allDone() {
-    endTime = time.milliseconds
```
Comment (afine) [inline core/src/main/scala/kafka/log/LogCleaner.scala:666]: changed from absolute to relative time

**Comment by gwenshap** [comment]
In addition to nanotime being relative, it is also much slower (http://stackoverflow.com/questions/19052316/why-is-system-nanotime-way-slower-in-performance-than-system-currenttimemill) and more precise (but not necessarily more accurate.

Since it is much slower I suggest not having the millisecond-level relativeTime - there is no reason to pay additional performance cost for no extra precision. It is seriously misleading (and perhaps you even mislead yourself...)

Regarding whether the other names make sense, I'll defer to @ijuma who created the original JIRA.

**Comment by ijuma** [comment]
One other advantage of System.nanoTime is that it is monotonically increasing. With regards to the performance cost, here's a thorough and reliable analysis on the cost of `System.nanoTime`:

http://shipilev.net/blog/2014/nanotrusting-nanotime/

I was unable to find an analysis of similar quality for `System.currentTimeMilllis`. We should definitely quantify the performance difference (particularly on Linux) before we choose one way or another.

Also, it would be good to do KAFKA-2247 before we do this.

**Comment by gwenshap** [comment]
Quantifying the performance differences will be great, but regardless - there is an important qualitative difference - on our target OS (Linux and Windows), currentTimeMillis does not require an interrupt and nanoTime does. Interrupts are a whole different story on systems that are already running at capacity (since they cause the process to lose its place on CPU queue, which on congested systems can be a huge time sink). For this reason I think its important to avoid nanoTime when at all possible.

**Comment by ijuma** [comment]
@gwenshap Yes, the impact on machines running at capacity is very important indeed. Aleksey covered this in the blog post above and he called it the scalability challenge. It's a shame that the blog post from David Holmes (https://blogs.oracle.com/dholmes/entry/inside_the_hotspot_vm_clocks) is from 2006 and a bit Windows-centric. It would be quite handy to have updated blog post as many things have changed since then (at all levels, CPUs, OSes, JVMs).

**Comment by ijuma** [comment]
For Mac OS X, some numbers are available here: https://bugs.openjdk.java.net/browse/JDK-8040140?focusedCommentId=13484433&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13484433

System.nanoTime becomes more costly than System.currentTimeMillis as the number of threads increases.

**Comment by ijuma** [comment]
http://pzemtsov.github.io/2017/07/23/the-slow-currenttimemillis.html indicates that `nanoTime` and `currentTimeMillis` perform similarly on Linux. Given that, it seems like this PR should be OK (after rebase).

**Pull request closed without merging**